source/extensions/http/header_validators/envoy_default/path_normalizer.cc

```diff
@@ -0,0 +1,403 @@
+  }
+
+  absl::string_view normalized_path{path};
+  // Update the :path header. We need to honor the normalized path and the original query/fragment
+  // components.
+  header_map.setPath(absl::StrCat(normalized_path, query));
+
+  if (redirect) {
+    return {PathNormalizationResult::Action::Redirect,
+            PathNormalizerResponseCodeDetail::get().RedirectNormalized};
+  }
+
+  return PathNormalizationResult::success();
+}
+
+PathNormalizer::PathNormalizationResult PathNormalizer::decodePass(std::string& path) const {
+  auto begin = path.begin();
+  auto read = std::next(begin);
+  auto write = std::next(begin);
+  auto end = path.end();
+  bool redirect = false;
+
+  while (read != end) {
+    if (*read == '%') {
+      auto decode_result = normalizeAndDecodeOctet(read, end);
+      // TODO(#23885) - add and honor config to not reject invalid percent-encoded octets.
+      switch (decode_result.result()) {
+      case PercentDecodeResult::Invalid:
+        ABSL_FALLTHROUGH_INTENDED;
+      case PercentDecodeResult::Reject:
+        // Reject the request
+        return {PathNormalizationResult::Action::Reject, UhvResponseCodeDetail::get().InvalidUrl};
+
+      case PercentDecodeResult::Normalized:
+        // Valid encoding but outside the UNRESERVED character set. The encoding was normalized to
+        // UPPERCASE and the octet must not be decoded. Copy the normalized encoding.
+        *write++ = *read++;
+        *write++ = *read++;
+        *write++ = *read++;
+        break;
+
+      case PercentDecodeResult::DecodedRedirect:
+        // The encoding was properly decoded but, based on the config, the request should be
+        // redirected to the normalized path.
+        redirect = true;
+        ABSL_FALLTHROUGH_INTENDED;
+      case PercentDecodeResult::Decoded:
+        // The encoding was decoded. Store the decoded octet in the last character of the percent
+        // encoding (read[2]) so it will be processed in the next iteration. We can safely advance
+        // 2 positions since we know that the value was correctly decoded.
+        std::advance(read, 2);
+        *read = decode_result.octet();
+      }
+    } else {
+      *write++ = *read++;
+    }
+  }
+
+  path.resize(std::distance(begin, write));
+  if (redirect) {
+    return {PathNormalizationResult::Action::Redirect,
+            PathNormalizerResponseCodeDetail::get().RedirectNormalized};
+  }
+
+  return PathNormalizationResult::success();
+}
+
+PathNormalizer::PathNormalizationResult PathNormalizer::mergeSlashesPass(std::string& path) const {
+  auto begin = path.begin();
+  auto read = std::next(begin);
+  auto write = std::next(begin);
+  auto end = path.end();
+
+  while (read != end) {
+    if (*read == '/') {
+      char prev = *std::prev(write);
+      if (prev == '/') {
+        // Duplicate slash, merge it
+        ++read;
+      } else {
+        // Not a duplicate slash or we aren't configured to merge slashes, copy it
```

Comment:
I don't see the code that handles the "... we aren't configured to merge slashes". I guess the comment should be updated.